Mathlib/Algebra/Module/Presentation/Basic.lean

```diff
@@ -0,0 +1,284 @@
+  var g := π (Finsupp.single g 1)
+  relation r := by
+    have : π = Finsupp.lift _ _ _ (fun g ↦ π (Finsupp.single g 1)) := by ext; simp
+    rw [← this]
+    exact hπ r
+
+@[simp]
+lemma ofπ_π : (ofπ π hπ).π = π := by ext; simp [ofπ]
+
+end
+
+section
+
+variable (π : (relations.G →₀ A) →ₗ[A] M) (hπ : π.comp relations.map = 0)
+
+/-- Variant of `ofπ` where the vanishing condition is expressed in terms
+of a composition of linear maps. -/
+noncomputable def ofπ' : relations.Solution M :=
+  ofπ π (fun r ↦ by
+    simpa using DFunLike.congr_fun hπ (Finsupp.single r 1))
+
+@[simp]
+lemma ofπ'_π : (ofπ' π hπ).π = π := by simp [ofπ']
+
+end
+
+/-- Given `relations : Relations A`, an `A`-module `M` and `solution : relations.Solution M`,
+this property asserts that `solution` gives a presentation of `M` by generators and relations. -/
+structure IsPresentation (solution : relations.Solution M) : Prop where
+  bijective : Function.Bijective solution.fromQuotient
+
+namespace IsPresentation
+
+variable {solution : relations.Solution M} (h : solution.IsPresentation)
+
+/-- When `M` admits a presentation by generators and relations given
+by `solution : relations.Solutions M`, this is the associated linear equivalence
+`relations.Quotient ≃ₗ[A] M`. -/
+noncomputable def linearEquiv : relations.Quotient ≃ₗ[A] M := LinearEquiv.ofBijective _ h.bijective
+
+@[simp]
+lemma linearEquiv_apply (x : relations.Quotient) :
+    h.linearEquiv x = solution.fromQuotient x := rfl
+
+@[simp]
+lemma linearEquiv_symm_var (g : relations.G) :
+    h.linearEquiv.symm (solution.var g) = Submodule.Quotient.mk (Finsupp.single g 1) :=
+  h.linearEquiv.injective (by simp)
+
+variable {N : Type v'} [AddCommGroup N] [Module A N]
+
+/-- If `M` admits a presentation by generators and relations, and we have a solution of the
+same equations in a module `N`, then this is the canonical induced linear map `M →ₗ[A] N`. -/
+noncomputable def desc (s : relations.Solution N) : M →ₗ[A] N :=
+  s.fromQuotient.comp h.linearEquiv.symm.toLinearMap
+
+@[simp]
+lemma desc_var (s : relations.Solution N) (g : relations.G) :
+    h.desc s (solution.var g) = s.var g := by
+  dsimp [desc]
+  rw [linearEquiv_symm_var, fromQuotient_single]
+
+include h in
+lemma postcomp_injective {f f' : M →ₗ[A] N}
+    (h' : solution.postcomp f = solution.postcomp f') : f = f' := by
+  suffices f.comp solution.fromQuotient = f'.comp solution.fromQuotient by
+    ext x
+    obtain ⟨y, rfl⟩ := h.bijective.2 x
+    exact DFunLike.congr_fun this y
+  ext g
+  simpa using congr_fun (congr_arg Solution.var h') g
+
+/-- If `M` admits a presentation by generators and relations, then
+linear maps from `M` can be (naturally) identified to the solutions of
+certain linear equations. -/
+@[simps]
+noncomputable def linearMapEquiv : (M →ₗ[A] N) ≃ relations.Solution N where
+  toFun f := solution.postcomp f
+  invFun s := h.desc s
+  left_inv f := h.postcomp_injective (by aesop)
+  right_inv s := by aesop
+
```

Comment:
Can we get the exact sequence $R^R \to R^G \to M \to 0$ given an `IsPresentation`?

Reply:
Thanks for the suggestion. I have just added this.